File: src/main/java/br/com/blz/testjava/infrastracture/product/persistence/ProductRepository.java
package br.com.blz.testjava.infrastracture.product.persistence;

import br.com.blz.testjava.domain.product.Sku;
import br.com.blz.testjava.infrastracture.exceptions.ConstraintViolationExceptionException;
import org.springframework.stereotype.Component;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;

@Component
public class ProductRepository {

    private final Map<Long, ProductEntity> products;

    public ProductRepository() {
        this.products = new LinkedHashMap<>();
    }

    public ProductEntity save(ProductEntity productEntity) {
        if (products.containsKey(productEntity.getSku())) {
            throw ConstraintViolationExceptionException.with("Produto", Sku.from(productEntity.getSku()), "sku");
        }
        products.put(productEntity.getSku(), productEntity);
        return productEntity;
    }

    public ProductEntity update(ProductEntity productEntity) {
        products.put(productEntity.getSku(), productEntity);
        return productEntity;
    }

    public Optional<ProductEntity> findBySku(Long sku) {
        return Optional.ofNullable(products.get(sku));
    }

    public void delete(Long sku) {
        products.remove(sku);
    }

    public Integer count() {
        return products.size();
    }
}

File: src/main/kotlin/br/com/blz/testkotlin/repository/ProductRepository.kt
package br.com.blz.testkotlin.repository

import br.com.blz.testkotlin.entity.ProductEntity
import org.springframework.stereotype.Repository

@Repository
class ProductRepository {

  var products = mutableListOf<ProductEntity>()

  fun save(productEntity: ProductEntity): Boolean {
    return products.add(productEntity)
  }

  fun findBySku(sku: Long): ProductEntity? {
    return products.find { it.sku == sku }
  }

  fun delete(sku: Long): Boolean {
    return products.removeIf { it.sku == sku }
  }

  fun getAll(): MutableList<ProductEntity>? {
    return products
  }

  fun update(productEntity: ProductEntity) {
    findBySku(productEntity.sku)?.apply {
      this.name = productEntity.name
      this.inventory.warehouses = productEntity.inventory.warehouses
    }
  }
}
